### PR TITLE
[release branch] better auto-generated agent docs

### PR DESCRIPTION
**Pull request by rwcarlsen**
Sample xml input snippets are generated for each state variable.  The header
for optional state variables now explicitly says:

    optional (default=[defualt])

Here is a screenshot:

snapshot1

### CODE REVIEW

File: source/cyclusagent.py [hunk 1]
```diff
@@ -85,6 +86,142 @@ def nicestr(x):
...
+    elif t == 'std::map':
+        name = 'map'
+        if isinstance(names[0], STRING_TYPES):
+            names[0] = [names[0], None]
+        elif names[0] is None:
+            names[0] = [name, None]
+        if names[0][0] is not None:
+            name = names[0][0]
+        itemname ='item' if names[0][1] is None else names[0][1]
+        keynames = 'key' if isinstance(cpptype[1], STRING_TYPES) else ['key']
```
Comment (mbmcgarry) [inline source/cyclusagent.py]: missing space after =

File: source/cyclusagent.py [hunk 2]
```diff
@@ -85,6 +86,142 @@ def nicestr(x):
...
+        impl += '</{0}>'.format(name)
+    elif t == 'std::map':
+        name = 'map'
+        if isinstance(names[0], STRING_TYPES):
+            names[0] = [names[0], None]
+        elif names[0] is None:
+            names[0] = [name, None]
+        if names[0][0] is not None:
+            name = names[0][0]
+        itemname ='item' if names[0][1] is None else names[0][1]
```
Comment (mbmcgarry) [inline source/cyclusagent.py]: missing space after =

File: source/cyclusagent.py [hunk 3]
```diff
@@ -181,6 +330,18 @@ def append_statevars(self):
...
 
+            self.lines += ['', ind + '.. code-block:: xml', '']
+            schemalines = buildsample(t, schematype, uitype, labels).split('\n')
+            previndent = ''
+            for l in schemalines:
+                if len(l.strip()) > 0:
+                    if l.strip() == '...':
+                        l = previndent + l.strip()
+                    self.lines.append(ind + '    ' + l)
+                    previndent = ' '*(len(l) - len(l.lstrip()))
```
Comment (mbmcgarry) [inline source/cyclusagent.py]: should there be spaces around * ?

File: source/cyclusagent.py (hunk 2 above)
Comment (rwcarlsen) [inline source/cyclusagent.py]: done.

File: source/cyclusagent.py (hunk 3 above)
Comment (rwcarlsen) [inline source/cyclusagent.py]: done.

**Comment by gidden** [comment]
I like it, thanks @rwcarlsen

File: source/cyclusagent.py [hunk 4]
```diff
@@ -85,6 +86,142 @@ def nicestr(x):
...
 
+def prepare_type(cpptype, othertype):
+    """Updates othertype to conform to the length of cpptype using None's.
+    """
+    if not isinstance(cpptype, STRING_TYPES):
+        if isinstance(othertype, STRING_TYPES):
+            othertype = [othertype]
+
+        if othertype is None:
+            othertype = [None] * len(cpptype)
```
Comment (gidden) [inline source/cyclusagent.py]: `othertype = othertype or [None] * len(cpptype)`

Comment (gidden) [inline source/cyclusagent.py]: nm, still in the conditional block

**Comment by rwcarlsen** [comment]
Note that this may need to change a bit depending on the fate of cyclus/cyclus#1143

**Comment by gidden** [comment]
Ok, do you mind if we keep it in the hopper until that's resolved?

On Wed, May 6, 2015 at 1:41 PM, Robert Carlsen <notifications@github.com>
wrote:

> Note that this may need to change a bit depending on the fate of
> cyclus/cyclus#1143 <https://github.com/cyclus/cyclus/pull/1143>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cyclus/cyclus.github.com/pull/139#issuecomment-99569085>
> .
>



-- 
Matthew Gidden, Ph.D.
Postdoctoral Associate, Nuclear Engineering
The University of Wisconsin -- Madison
Ph. 225.892.3192

**Comment by rwcarlsen** [comment]
fine with me.

**Comment by mbmcgarry** [comment]
Incidentally, I also think this PR will be a GREAT addition to the documentation.  Nice idea @rwcarlsen  :)

**Comment by rwcarlsen** [comment]
Since cyclus/cyclus#1143 went in, this doesn't need any changes and should be ready for a merge.

**Comment by gidden** [comment]
thanks @rwcarlsen

**Pull request merged**